### ISSUE TITLE
error: Panic if no syslog socket exists.

### ISSUE THREAD

**Issue by EStrehlau** [usbarmory/crucible#17]
Hello, I am using the -s option of crucible to get machine readable output during early boot.

This is before the syslog daemon has been started and can not be shifted to a later point.

Sadly crucible panics in that case:
```
Sat Jan  1 00:21:28 2000: panic: runtime error: invalid memory address or nil pointer dereference
Sat Jan  1 00:21:28 2000: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x107abc]
Sat Jan  1 00:21:28 2000:
Sat Jan  1 00:21:28 2000: goroutine 1 [running]:
Sat Jan  1 00:21:28 2000: log/syslog.(*Writer).Write(0x0, {0x1cd01e0, 0x4d, 0x50})
Sat Jan  1 00:21:28 2000:       log/syslog/syslog.go:178 +0x14
Sat Jan  1 00:21:28 2000: log.(*Logger).Output(0x1c80060, 0x2, {0x1cd0190, 0x4c})
Sat Jan  1 00:21:28 2000:       log/log.go:195 +0x2bc
Sat Jan  1 00:21:28 2000: log.Printf({0x19acb7, 0xb}, {0x1cc9f0c, 0x3, 0x3})
Sat Jan  1 00:21:28 2000:       log/log.go:351 +0x8c
Sat Jan  1 00:21:28 2000: main.read({0x1ce0180, 0x26}, 0x1c802a0, {0xbe8c8fdf, 0xa})
Sat Jan  1 00:21:28 2000:       github.com/usbarmory/crucible/cmd/crucible/crucible.go:257 +0x5b0
Sat Jan  1 00:21:28 2000: main.main()
Sat Jan  1 00:21:28 2000:       github.com/usbarmory/crucible/cmd/crucible/crucible.go:389 +0x79c
Sat Jan  1 00:21:28 2000: error executing crucible
Sat Jan  1 00:21:28 2000: panic: runtime error: invalid memory address or nil pointer dereference
Sat Jan  1 00:21:28 2000: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x107abc]
Sat Jan  1 00:21:28 2000:
Sat Jan  1 00:21:28 2000: goroutine 1 [running]:
Sat Jan  1 00:21:28 2000: log/syslog.(*Writer).Write(0x0, {0x1d0e1e0, 0x4d, 0x50})
Sat Jan  1 00:21:28 2000:       log/syslog/syslog.go:178 +0x14
Sat Jan  1 00:21:28 2000: log.(*Logger).Output(0x1c50060, 0x2, {0x1d0e190, 0x4c})
Sat Jan  1 00:21:28 2000:       log/log.go:195 +0x2bc
Sat Jan  1 00:21:28 2000: log.Printf({0x19acb7, 0xb}, {0x1d07f0c, 0x3, 0x3})
Sat Jan  1 00:21:28 2000:       log/log.go:351 +0x8c
Sat Jan  1 00:21:28 2000: main.read({0x1d1e180, 0x26}, 0x1c502a0, {0xbeabcfdf, 0xa})
Sat Jan  1 00:21:28 2000:       github.com/usbarmory/crucible/cmd/crucible/crucible.go:257 +0x5b0
Sat Jan  1 00:21:28 2000: main.main()
Sat Jan  1 00:21:28 2000:       github.com/usbarmory/crucible/cmd/crucible/crucible.go:389 +0x79c
Sat Jan  1 00:21:28 2000: error executing crucible
```

I think that machine readable output and log location should be seperate concerns. It would be nice if there was a flag to toggle between syslog/stderr and one to toggle between machine and human.

Nevertheless I think crashing in this case is a bug that needs fixing for my usecase.
In this case there should probably be no logging done.

I would submit a pull request for my flag proposal but I am not sure how to include this in a backwards compatible way.
I think keeping the -s flag to default to machine readable + syslog and then including another flag to force stderr is probably the best approach.

**Comment by abarisani** [comment]
Which Go version are you compiling Crucible with?

**Comment by EStrehlau** [comment]
I am using the 2022.05 binary release.

I think the issue should be a missing error check in:
https://github.com/usbarmory/crucible/blob/5e0805300e15bdebb8d50ea8bf1d1e2ac4f9dcb7/cmd/crucible/crucible.go#L334

Syslog.new delegates to:
https://cs.opensource.google/go/go/+/master:src/log/syslog/syslog_unix.go;drc=f229e7031a6efb2f23241b5da000c3b3203081d6;l=17
Which returns an error if the socket does not exist yet.

**Comment by abarisani** [comment]
If logwriter is nil (passed to log.SetOutput) I'd not expect Go to panic, it feels to me that Go log package should do better. Anyway I can avoid setting log output if logwriter is nil.

This [commit](https://github.com/usbarmory/crucible/commit/fa31c83dbab2e6fc5f5e5d4ef45f5c04ba084da2) should fix the issue, please close if confirmed. Thanks!

**Comment by EStrehlau** [comment]
Thanks I'll try and come back to you later today, looks like the default is logging to stderr so it should work out.

Sadly our toolchain (yocto dunfell release) does not support the version of go needed for crucible.

That is the reason I was using the binary release.
If it is not to much to ask could you do a new release that includes this fix?
Otherwise we have to include a manually build binary without provenance back to the official releases.

**Comment by abarisani** [comment]
So do you consider this fixed or not? If not which behavior would you like to have? I can just bail if -s is set and no logger is available.

**Comment by EStrehlau** [comment]
The problem is we are logging to stdout and not stderr.
I am using the -s option to get machine readable output, if we log to stdout tho when -s is passed the output is interspersed with the log messages.

In other words, there currently is no way to get machine readable output if syslog is not installed.

I think it would make sense to log to stderr by default in case -s is set and upgrade to syslog if it is available.

Otherwise the -s option is still useless if syslog is not around sadly.

**Comment by abarisani** [comment]
Why can't you redirect stdout to a file of your choice? I don't understand why stderr, instead of stdout, would make things easier for you.

**Comment by EStrehlau** [comment]
Because we have two streams, the machine readable output and the log.

Currently, if syslog is found all logging goes to syslog and all machine readable output to stdout.

This is perfect since stdout ONLY contains machine readable output, that can be directly parsed.

With the change as is the log and the supposedly machine readable output get "mixed together" in stdout if there is no syslog.
So stdout contains machine readable output AND logging messages in case syslog is not found.

This makes it useless for interpreting the output since now you all of the sudden need to filter out the log messages.

That is why unix utilities usually write output for human consumption to stderr and for machine consumption to stdout.

Sorry that I wasn't clear enough on that part before.

Basically I need logging and machine output on two different streams like it is when syslog is found.
I don't really care about the logging at all, it could be disabled for my usecase. What I do care about is getting a stream of only machine readable output.

**Comment by abarisani** [comment]
Please see if this would work:

```
diff --git a/cmd/crucible/crucible.go b/cmd/crucible/crucible.go
index eea54e4..e9aae10 100644
--- a/cmd/crucible/crucible.go
+++ b/cmd/crucible/crucible.go
@@ -333,6 +333,8 @@ func main() {
 	if conf.syslog {
 		if logwriter, _ := syslog.New(syslog.LOG_INFO, "crucible"); logwriter != nil {
 			log.SetOutput(logwriter)
+		} else {
+			log.SetOutput(os.Stderr)
 		}
 	} else {
 		log.SetOutput(os.Stdout)
```

**Comment by EStrehlau** [comment]
Working with your patch now.

Thank you so much!

It would be great if you could do a release so we could actually use the fix.

**Comment by andrejro** [comment]
New release containing the fix has been created: https://github.com/usbarmory/crucible/releases/tag/v2023.04.12
